### PR TITLE
Invisible Lasers

### PR DESCRIPTION
**Pull request by JustWannaFly**
[Invisible Lasers](https://github.com/BuildCraft/BuildCraft/issues/3513) implemented via configuration option. Currently in the experimental configuration category, but that is subject to change once a proper place is figured out and the ability to view invisible lasers with goggles is implemented

### CODE REVIEW

**Comment by AEnterprise** [comment]
travis failure is again travis killing the build for some reason, gona have to investigate it
you might also want to link that email to your github (or use your github email for future commits)

File: common/buildcraft/silicon/BCSiliconConfig.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        Configuration config = BCCoreConfig.config;
+        //TODO change the category and update the comment once visible with goggles is implemented
+        propRenderLaserBeams = config.get("experimental", "renderLaserBeams", true,
+                "When false laser beams will not be visible while transmitting power");
+        EnumRestartRequirement.WORLD.setTo(propRenderLaserBeams);
+
+        MinecraftForge.EVENT_BUS.register(BCSiliconConfig.class);
+    }
+
+    public static void reloadConfig(EnumRestartRequirement restarted) {
```
Comment (AEnterprise) [inline common/buildcraft/silicon/BCSiliconConfig.java]: the restarted param is not needed, it's not used at all

File: common/buildcraft/silicon/BCSiliconConfig.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        renderLaserBeams = propRenderLaserBeams.getBoolean();
+    }
+
+    @SubscribeEvent
+    public static void onConfigChange(OnConfigChangedEvent cce) {
+        if (BCModules.isBcMod(cce.getModID())) {
+            EnumRestartRequirement req = EnumRestartRequirement.NONE;
```
Comment (AEnterprise) [inline common/buildcraft/silicon/BCSiliconConfig.java]: the restart requirement should always be NONE, it is applied instantly and we don't need a world reload to start/stop rendering

File: common/buildcraft/silicon/BCSiliconConfig.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+    private static Property propRenderLaserBeams;
+
+    public static void preInit() {
+
+        Configuration config = BCCoreConfig.config;
+        //TODO change the category and update the comment once visible with goggles is implemented
+        propRenderLaserBeams = config.get("experimental", "renderLaserBeams", true,
+                "When false laser beams will not be visible while transmitting power");
+        EnumRestartRequirement.WORLD.setTo(propRenderLaserBeams);
```
Comment (AEnterprise) [inline common/buildcraft/silicon/BCSiliconConfig.java]: See below, the requirement should always be NONE, just set it here and leave it at none then

File: common/buildcraft/silicon/BCSilicon.java [hunk 4]
```diff
@@ -40,9 +41,13 @@
     public void preInit(FMLPreInitializationEvent evt) {
         RegistryHelper.useOtherModConfigFor(MODID, BCCore.MODID);
 
+        BCSiliconConfig.preInit();
         BCSiliconItems.preInit();
         BCSiliconBlocks.preInit();
 
+        // Reload after all of the definitions have been created.
+        BCSiliconConfig.reloadConfig(EnumRestartRequirement.GAME);
```
Comment (AlexIIL) [inline common/buildcraft/silicon/BCSilicon.java]: Put this in the config preinit please

File: common/buildcraft/silicon/BCSiliconConfig.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+import net.minecraftforge.common.config.Configuration;
+import net.minecraftforge.common.config.Property;
+import net.minecraftforge.fml.client.event.ConfigChangedEvent.OnConfigChangedEvent;
+import net.minecraftforge.fml.common.Loader;
+import net.minecraftforge.fml.common.LoaderState;
+import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
+
+public class BCSiliconConfig {
+
+    public static Boolean renderLaserBeams = true;
```
Comment (AlexIIL) [inline common/buildcraft/silicon/BCSiliconConfig.java]: Bit minor, but this should be a primitive, not the object type.

File: common/buildcraft/silicon/BCSiliconConfig.java (hunk 3 above)
Comment (AlexIIL) [inline common/buildcraft/silicon/BCSiliconConfig.java]: You can just remove this line to make it work.

**Comment by AlexIIL** [comment]
Thanks! As for goggles, the item can be found in BCCoreItems.goggles, and you can probably add a check against Minecraft.gteMinecraft().getPlayer(), and see if the item equipped on their head equals the goggles.

**Comment by AEnterprise** [comment]
@AlexIIL the restart params sound like something that isn't needed nor usefull at all, they are only used to tell forge what needs to be restarted, it shoudn't be used at all to tell a config what reloaded, they shoudn't care about that at all

File: common/buildcraft/silicon/BCSiliconConfig.java (hunk 5 above)
Comment (JustWannaFly) [inline common/buildcraft/silicon/BCSiliconConfig.java]: Apparently I've been doing too much JavaScript at work lately if I can't remember how to properly initialize a typed variable :stuck_out_tongue_winking_eye:

**Comment by AlexIIL** [comment]
@AEnterprise the reload parameter is used both to tell forges gui what should be enabled/disabled in the gui, and so that Buildcraft only changes the values that can actually be changed during the cycle.

@JustWannaFly please leave the reload parameter as-is, it's simpler if all config should have the same method signature (as then we can move the onConfig changed event to lib and add the methods to a single list for reloading.

**Comment by AEnterprise** [comment]
@AlexIIL that is not how configs or properties work at all

if we set a property to require a world restart then that property cannot be eddited ingame, you need to leave the world and edit from the main menu to be able to edit those. So they can safely be updated as the only time they are actually changed is when it's safe to edit those

as for configs requiring a minecraft restart those are also not editable ingame, excluding those from being reloaded also doesn't make sense because those are values we only need during startup so changing them doesn't hurt anything (and forge makes the player click 'i understand' about the needing a restart)

**Comment by JustWannaFly** [comment]
I'm seeing the build fail with the below message:

:extractRangemapReplacedMain
/home/travis/build/BuildCraft/BuildCraft/build/sources/main/java
/home/travis/.travis/job_stages: line 57:  3702 Killed                  ./gradlew build
The command "./gradlew build" exited with 137.
Done. Your build exited with 1.

Any insight into what causes this? My googling just reveals potential issues with running out of memory or needing the travis set up in a container. I admit I don't know anything about travis though.

File: common/buildcraft/silicon/client/render/RenderLaser.java [hunk 6]
```diff
@@ -57,4 +59,15 @@ public void renderTileEntityFast(@Nonnull TileLaser tile, double x, double y, do
             Minecraft.getMinecraft().mcProfiler.endSection();
         }
     }
+
+    private boolean isPlayerWearingBCGoggles () {
```
Comment (AlexIIL) [inline common/buildcraft/silicon/client/render/RenderLaser.java]: Just `isPlayerWearingGoggles`, or lowercase `Bc`

File: common/buildcraft/silicon/BCSiliconConfig.java [hunk 7]
```diff
@@ -24,9 +24,8 @@
     public static void preInit() {
 
         Configuration config = BCCoreConfig.config;
-        //TODO update the comment once visible with goggles is implemented
         propRenderLaserBeams = config.get("display", "renderLaserBeams", true,
-                "When false laser beams will not be visible while transmitting power");
+                "When false laser beams will not be visible while transmitting power without wearing the BC Goggles");
```
Comment (AlexIIL) [inline common/buildcraft/silicon/BCSiliconConfig.java]: Please use the full name (`BuildCraft`) or don't, rather than `BC`

**Comment by AlexIIL** [comment]
Ok, the memory issue is fixed on the 8.0.x branch, so can you merge and push?

**Comment by JustWannaFly** [comment]
Switched to just checking the head slot instead of iterating through all the armor slots. Definitely cleaner, but this is the first mod I've worked on so please keep letting me know if I'm doing anything silly that can be done a lot simpler.

**Comment by JustWannaFly** [comment]
My apologies for the multiple failed builds. It took me a bit to get Intellij configured properly. All should be good now

**Comment by AlexIIL** [comment]
Looks good, thanks :)

I think there are a few things that need changing about the config system, but some other classes want changing as well so I'll fix them later.

**Pull request merged**